sourcegraph/cody
# feat: code smell command, display commands error in chat

RE feedback from beyang:

- Could we add tab-to-complete behavior, so if the user has typed in a partial match for the command name (e.g., “/expla”) and then hits Tab, we complete the command name (in this case “/explain”)?
- I noticed nothing happens if I execute /explain without anything selected. It seems intuitive to me to use the currently open file as context in this case.
- I think it makes sense to display any errors (like you need to have something selected) as a response in the chat, rather than as a separate error pop-up

This PR includes the following change:
1. for the default `/explain` & `/smell` command, run on visible content when there is no selection in the current text document
1. Display error message from commands in chat as an error message from the assistant when selection is required for a command (e.g. /doc and /test) but currently missing:
1. log basic telemetryService for commands
1. also cleaned up the custom-prompt recipe that all commands depend on.
1. show error for invalid commands:
<img width="515" alt="image" src="https://github.com/sourcegraph/cody/assets/68532117/85e86e59-916c-465e-9fd0-3ce4098e8dc3">
1. add `/smell`, command for code smell as the default command
1. display the file name and selection range when available
<img width="1363" alt="image" src="https://github.com/sourcegraph/cody/assets/68532117/0d643db7-0499-4c9f-bef7-dec963c8f5c2">

#### Some UI changes were moved to separate PRs

See moved to https://github.com/sourcegraph/cody/pull/606 & https://github.com/sourcegraph/cody/pull/648
- add tab-to-complete +  enter-to-complete behavior in the command list
- update the walkthrough to replace recipes with commands and add a new commands section:
- add short instructions on how to use commands with the new menu. The link also works as a button to open the command menu:
- update shortcut for commands menu to `alt+c` as it was using the same key as `cody.fixup.new`

## Test plan



See the complete demo:


https://github.com/sourcegraph/cody/assets/68532117/64f18b91-1fa4-4bfa-ae8a-acbce6360a39

1. start cody from this branch
2. run `/doc` without code selection - error expected
3. run non-exisit command:`/exit` - error expected
4. run `/explain` and `smell` without code selection - both show work with visible content
5. clicking on the `Cody Menu`button in the cody intro should open the command menu for you
7. shortcut `alt+c` should bring up the command menu
8. run commands from the right-click menu, everything should work accordingly
